### PR TITLE
Add FileSystem proposed protocol spec & documentation

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added the spec itself in addition to adding the protocol types, requests, notifications and expansions of that in the API.
- Prior to updating implementation figured this was a good first step to ensure I was setting things up correctly.

### CODE REVIEW

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+ * updated `mtime` that advanced from the previous value in the [stat](#FileStat) and a
+ * correct `size` value. Otherwise there may be optimizations in place that will not show
+ * the change in an editor for example.
+ *
+ * @since 3.17.0 - proposed state
+ */
+export interface DidChangeFileParams {
+	/**
+     * The change events.'
+     */
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:173]:
```suggestion
	/**
	 * The change events.
	 */
```

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+/**
+ * The event filesystem providers must use to signal a file change.
+ *
+ * @since 3.17.0 - proposed state
+ */
+export interface FileChangeEvent {
+	/**
+     * The type of change.
+     */
+	uri: string;
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:186]: I would suggest changing any `uri: string` declarations into `uri: DocumentUri`.

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+ */
+export interface DidChangeFileParams {
+	/**
+     * The change events.'
+     */
+	changes: FileChangeEvent[];
+}
+
+/**
+ * The event filesystem providers must use to signal a file change.
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:178]: There's a wording issue here. Seems like a noun should follow "must use"?

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+
+	/**
+     * A file has been created.
+     */
+	Created = 2,
+
+	/**
+     * A file has been deleted.
+     */
+	Deleted = 3,
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:213]: The formatting is inconsistent here.

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+
+	/**
+     * The subscription ID to be used in order to stop watching the provided file or folder uri via the fileSystem/stopWatching notification.
+     */
+	subscriptionId: string;
+
+	/**
+     * Configures the watch
+     */
+	options: WatchFileOptions
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:249]: Some formatting issues here also.

File: protocol/src/common/proposed.fileSystemProvider.ts [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+	mtime: number;
+
+	/**
+     * The size in bytes.
+     *
+     * *Note:* If the file changed, it is important to provide an updated `size`. Otherwise there
+     * may be optimizations in place that will not show the updated file contents in an editor for
+     * example.
+     */
+	size: number;
```
Comment (rcjsuen) [inline protocol/src/common/proposed.fileSystemProvider.ts:355]: Perhaps `unumber` can be used here instead to indicate that these must be non-negative numbers.

**Review by rcjsuen — COMMENTED** [review]
Looking at the pull request on GitHub I can see a lot formatting inconsistencies.

**Comment by dbaeumer** [comment]
Closing the PR. If we want to add this I think we have to start fresh.

**Pull request closed without merging**